guide/oauth2/README.md

```diff
@@ -0,0 +1,200 @@
+
+After you change the response type, you can test the url right away. Try visiting it in your browser and you will be directed to a page that looks like this.
+
+![img](/assets/img/49jali8.png)
+
+You can see that by clicking `Authorize`, you are allowing the application to access your username and avatar. Once you click through, you should be redirected to the redirect url with `#access_token=ACCESS_TOKEN&token_type=Bearer&expires_in=EXPIRATION&scope=identify` appended to it. You now have an access token and can make requests to Discord's API to get information on the user. Let's modify `index.html` to add our OAuth2 url and to take advantage of the access token if it exists.
+
+```html
+<!DOCTYPE html>
+<div id='info'>
+	Hoi!
+</div>
+<a id='login' style='display:none;' href='your oauth2 url here'>Identify Yourself</a>
+<script>
+	window.onload = () => {
+		const match = window.location.hash.match(/access_token=(.+?)&token_type=(.+?)(?:&|$)/);
+		if (match) {
+			const [, access_token, token_type] = match;
+			const xml = new XMLHttpRequest();
+			xml.onload = () => {
+				const { username, discriminator } = JSON.parse(xml.response);
+				document.getElementById('info').innerText += `, ${username}#${discriminator}`;
+			}
+			xml.open('GET', 'https://discordapp.com/api/users/@me');   
+			xml.setRequestHeader('authorization', token_type + ' ' + access_token);
+			xml.send();
+		} else {
+			document.getElementById('login').style.display = 'block';
+		}
+	}
+</script>
+```
+
+Here, we just grab the access token and type from the url if it's there and use it to get info on the user, which is then used to greet them. In the following sections, we'll go over various details of Discord and OAuth2.
+
+## More details
+
+### The state parameter 
+
+OAuth2's protocols provide a `state` parameter which is supported by Discord. This is used to help prevent [CSRF](https://en.wikipedia.org/wiki/Cross-site_request_forgery) attacks and can also be used to represent the state of your application. This should be generated per user and appended to the OAuth2 url. For a very basic example, we can use a randomly generated string encoded in Base64 as the state parameter.
+
+```js
+function generateRandomString() {
+	const rand = Math.floor(Math.random() * 10);
+	let randStr = '';
+	for (let i = 0; i < 20 + rand; i++) {
+		randStr += String.fromCharCode(33 + Math.floor(Math.random() * 94));
+	}
+	return randStr;
+}
+
+// ...
+
+// generate and store the string
+const randStr = generateRandomString();
+localStorage.setItem('stateParameter', randStr);
+oauthURL += `&state=${btoa(randStr)}`;
```

Comment:
`oauthURL` is currently undefined as of right here. If someone tries to follow the guide, reading all the texts and using the code from the codeblocks and then get an error when they run the script, it seems a bit misleading/incomplete, which is why we try to avoid that in all the other pages. 

To avoid hardcoding the URL in a string or something, you can just do something like this for the next line:

```js
 document.getElementById('login').href += `&state=${btoa(randStr)}`;
```